### PR TITLE
Graphics/srp picking

### PR DESCRIPTION
**Pull request by joeante**
# **Please read;**
## **PR Workflow for the Graphics repository:**
* **All PRs must be opened as draft initially**
* Reviewers can be added while the PR is still in draft
* The PR can be marked as “Ready for Review” once the reviewers have confirmed that **no more changes are needed**
* Tests will start automatically after the PR is marked as “Ready for Review”
* **Do not use [skip ci]** - this can break some of our tooling
* Read the [Graphics repository & Yamato FAQ](http://go/graphics-yamato-faq).

### Checklist for PR maker
- [ ] Have you added a backport label (if needed)? For example, the `need-backport-*` label. After you backport the PR, the label changes to `backported-*`.
- [ ] Have you updated the changelog? Each package has a `CHANGELOG.md` file.
- [ ] Have you updated or added the documentation for your PR? When you add a new feature, change a property name, or change the behavior of a feature, it's best practice to include related documentation changes in the same PR.
- [ ] Have you added a graphic test for your PR (if needed)? When you add a new feature, or discover a bug that tests don't cover, please add a graphic test.

---
### Purpose of this PR
Why is this PR needed, what hard problem is it solving/fixing?

---
### Testing status

**Manual Tests**: What did you do?
- [ ] Opened test project + Run graphic tests locally
- [ ] Built a player
- [ ] Checked new UI names with UX convention
- [ ] Tested UI multi-edition + Undo/Redo + Prefab overrides + Alignment in Preset
- [ ] C# and shader warnings (supress shader cache to see them)
- [ ] Checked new resources path for the reloader (in developer mode, you have a button at end of resources that check the paths)
- Other: 

**Automated Tests**: What did you setup? (Add a screenshot or the reference image of the test please)

**Yamato**: (Select your branch):
https://yamato.prd.cds.internal.unity3d.com/jobs/902-Graphics

Any test projects to go with this to help reviewers?

---
### Comments to reviewers
Notes for the reviewers you have assigned.

### CODE REVIEW

File: com.unity.render-pipelines.universal/ShaderLibrary/Shadows.hlsl [hunk 1]
```diff
@@ -22,9 +22,7 @@
     #endif
 #endif
 
-#if defined(_ADDITIONAL_LIGHTS) || defined(_MAIN_LIGHT_SHADOWS_CASCADE)
-    #define REQUIRES_WORLD_SPACE_POS_INTERPOLATOR
-#endif
+#define REQUIRES_WORLD_SPACE_POS_INTERPOLATOR
```
Comment (joeante) [inline com.unity.render-pipelines.universal/ShaderLibrary/Shadows.hlsl:25]: Should be reverted?

File: com.unity.render-pipelines.universal/Editor/ShaderGraph/Targets/UniversalLitSubTarget.cs [hunk 2]
```diff
@@ -527,6 +559,56 @@ static class LitKeywords
                 { CoreKeywordDescriptors.ShadowsSoft },
                 { CoreKeywordDescriptors.MixedLightingSubtractive },
             };
+            
+            public static KeywordCollection DataExtraction = new KeywordCollection
```
Comment (joeante) [inline com.unity.render-pipelines.universal/Editor/ShaderGraph/Targets/UniversalLitSubTarget.cs:563]: no longer necessary?

File: com.unity.render-pipelines.universal/ShaderLibrary/Shadows.hlsl (hunk 1 above)
Comment (joeante) [reply com.unity.render-pipelines.universal/ShaderLibrary/Shadows.hlsl:25]: fixed

File: com.unity.render-pipelines.universal/Runtime/UniversalRenderPipeline.cs [hunk 3]
```diff
@@ -116,6 +117,113 @@ protected override void Dispose(bool disposing)
...
+
+                var data = ScriptableObject.CreateInstance<RenderRequestRendererData>();
+                data.request = renderRequest;
+                InitializeCameraData(camera, additionalCameraData, true, out var cameraData);
+                cameraData.renderer = data.InternalCreateRenderer();
+                RenderSingleCamera(renderContext, cameraData, false);
+                Object.DestroyImmediate(data);
+        }
+
+        public class RenderRequestRendererData : ScriptableRendererData
```
Comment (joeante) [inline com.unity.render-pipelines.universal/Runtime/UniversalRenderPipeline.cs:174]: Why is this public?

File: com.unity.shadergraph/Editor/Generation/Enumerations/KeywordType.cs [hunk 4]
```diff
@@ -4,6 +4,7 @@
     internal enum KeywordType
     {
         Boolean,
-        Enum
+        Enum,
+        MultiCompile
```
Comment (alindmanUnity) [inline com.unity.shadergraph/Editor/Generation/Enumerations/KeywordType.cs:8]: This is not necessary

File: com.unity.shadergraph/Editor/Data/Util/KeywordUtil.cs [hunk 5]
```diff
@@ -79,6 +79,10 @@ public static string ToDeclarationString(this KeywordDescriptor keyword)
                     var enumEntryDefinitions = keyword.entries.Select(x => $"{keyword.referenceName}_{x.referenceName}");
                     string enumEntriesString = string.Join(" ", enumEntryDefinitions);
                     return $"#pragma {definitionString} {enumEntriesString}";
+                case KeywordType.MultiCompile:
+                    var multiCompileEntryDefinitions = keyword.entries.Select(x => $"{x.referenceName}");
+                    string multiCompileEntriesString = string.Join(" ", multiCompileEntryDefinitions);
+                    return $"#pragma {definitionString} {multiCompileEntriesString}";
```
Comment (alindmanUnity) [inline com.unity.shadergraph/Editor/Data/Util/KeywordUtil.cs:85]: this is not necessary. the enum prefix (`keyword.referenceName`) is `RENDER` in this case. the existing enum keyword case already meets this need.

File: com.unity.render-pipelines.universal/Editor/ShaderGraph/Targets/UniversalLitSubTarget.cs [hunk 6]
```diff
@@ -527,6 +559,56 @@ static class LitKeywords
...
+            {
+                new KeywordDescriptor()
+                {
+                    displayName = "Extraction Modes",
+                    referenceName = "",
```
Comment (alindmanUnity) [inline com.unity.render-pipelines.universal/Editor/ShaderGraph/Targets/UniversalLitSubTarget.cs:568]: reference name should be `RENDER`

File: com.unity.render-pipelines.universal/Editor/ShaderGraph/Targets/UniversalLitSubTarget.cs [hunk 7]
```diff
@@ -527,6 +559,56 @@ static class LitKeywords
...
+                    type = KeywordType.MultiCompile,
```
Comment (alindmanUnity) [inline com.unity.render-pipelines.universal/Editor/ShaderGraph/Targets/UniversalLitSubTarget.cs:569]: type should be `Enum`

File: com.unity.render-pipelines.universal/Editor/ShaderGraph/Targets/UniversalLitSubTarget.cs [hunk 8]
```diff
@@ -527,6 +559,56 @@ static class LitKeywords
...
+                        new KeywordEntry("Object Id", "RENDER_OBJECT_ID"),
+                        new KeywordEntry("Depth", "RENDER_DEPTH"),
+                        new KeywordEntry("World Normals (face)", "RENDER_WORLD_NORMALS_FACE"),
+                        new KeywordEntry("World Normals (pixel)", "RENDER_WORLD_NORMALS_PIXEL"),
+                        new KeywordEntry("World Position", "RENDER_WORLD_POSITION"),
+                        new KeywordEntry("Base Color + Alpha", "RENDER_BASE_COLOR_ALPHA"),
+                        new KeywordEntry("Specular(RGB) Metallic(A)", "RENDER_SPECULAR_METALLIC"),
+                        new KeywordEntry("Emission(RGB)", "RENDER_EMISSION"),
+                        new KeywordEntry("Smoothness(r) + Occlusion (g)", "RENDER_SMOOTHNESS_OCCLUSION"),
+                        new KeywordEntry("EntityId", "RENDER_ENTITY_ID"),
```
Comment (alindmanUnity) [inline com.unity.render-pipelines.universal/Editor/ShaderGraph/Targets/UniversalLitSubTarget.cs:585]: each entry should have `RENDER_` removed from the reference name field, as the generator already formats these as needed for multicompile enum entries

File: com.unity.render-pipelines.universal/Editor/ShaderGraph/Targets/UniversalLitSubTarget.cs [hunk 9]
```diff
@@ -527,6 +559,56 @@ static class LitKeywords
...
+                        new KeywordEntry("UV0", "RENDER_SPACE_UV0"),
+                        new KeywordEntry("UV1", "RENDER_SPACE_UV1"),
+                        new KeywordEntry("UV2", "RENDER_SPACE_UV2"),
+                        new KeywordEntry("UV3", "RENDER_SPACE_UV3"),
+                        new KeywordEntry("UV4", "RENDER_SPACE_UV4"),
+                        new KeywordEntry("UV5", "RENDER_SPACE_UV5"),
+                        new KeywordEntry("UV6", "RENDER_SPACE_UV6"),
+                        new KeywordEntry("UV7", "RENDER_SPACE_UV7"),
+                        new KeywordEntry("UV8", "RENDER_SPACE_UV8"),
+                    },
```
Comment (alindmanUnity) [inline com.unity.render-pipelines.universal/Editor/ShaderGraph/Targets/UniversalLitSubTarget.cs:608]: same modifications as above

File: com.unity.render-pipelines.core/ShaderLibrary/DataExtraction.hlsl [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+#define RENDER_METALLIC_R              8
+#define RENDER_EMISSION_RGB            9
+#define RENDER_WORLD_NORMALS_PIXEL_RGB 10
+#define RENDER_SMOOTHNESS_R            11
+#define RENDER_OCCLUSION_R             12
+#define RENDER_DIFFUSE_COLOR_RGBA      13
+
+
+void ConvertSpecularToMetallic(float3 diffuseColor, float3 specularColor, out float3 baseColor, out float metallic)
+{
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.core/ShaderLibrary/DataExtraction.hlsl:21]: This file should be in universal. HDRP have already its own API (AOV API and Debug API) to extract information and will not rely on this shader.

Side question: Why do we have all the Specular color and Mettalic for SRP Picking. Guess there is a mix here of tooling helper for picking (object id, depth, normal, entity) and function to extract information for LOD. So we mix two different concept. Are we really sure we want to go through the same API?

Comment (joeante) [reply com.unity.render-pipelines.core/ShaderLibrary/DataExtraction.hlsl:21]: The main thing we need to solve is have a common API across render pipelines. An API that is specific to HDRP is not very useful when writing high level tools. Like Recorder / Imposter / LOD / Picking.

On the backend for HDRP it is perfectly fine if the API is just a wrapper around the AOV / Debug API.

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.cs [hunk 11]
```diff
@@ -3031,6 +3031,53 @@ static void BlitFinalCameraTexture(BlitFinalCameraTextureParameters parameters,
...
+            {
+                var aovRequest = AOVRequest.NewDefault(request.mode);
+                var aovBuffers = AOVBuffers.Color;
+                if (request.mode == Camera.RenderRequestMode.Depth || request.mode == Camera.RenderRequestMode.WorldPositionRGB)
+                    aovBuffers = AOVBuffers.Output;
+
+                //@TODO: _DebugBufferID handling seems dodgy what if the different buffers have different resolutions...
+                requests.Add(
+                    aovRequest,
+                    bufferId => _DebugBufferID ?? (_DebugBufferID = RTHandles.Alloc(request.result.width, request.result.height, 1, DepthBits.None, request.result.graphicsFormat)),
```
Comment (joeante) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.cs:3055]: use request.result directly

File: com.unity.render-pipelines.core/ShaderLibrary/DataExtraction.hlsl (hunk 10 above)
Comment (joeante) [reply com.unity.render-pipelines.core/ShaderLibrary/DataExtraction.hlsl]: Move to CommonMaterial.hlsl

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.cs (hunk 11 above)
Comment (joeante) [reply com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.cs]: done

**Comment by alindmanUnity** [comment]
What's the status of this PR? Do we still need to land this?

**Comment by phi-lira** [comment]
Any updates on this PR? Do we plan to work and land it?

**Comment by phi-lira** [comment]
Closing this PR as DOTS Picking PR is on process to land in SRP. https://github.com/Unity-Technologies/Graphics/pull/6071

**Pull request closed without merging**